### PR TITLE
Add ccagt dataset

### PR DESCRIPTION
**Pull request by johnnv1**
As described in #4075 

I could not generate the dummy data. Also, on the data repository isn't provided the split IDs, but I copy the functions that provide the correct data split. In summary, to have a better distribution, the data in this dataset should be separated based on the amount of NORs in each image.

### CODE REVIEW

**Comment by mariosasko** [comment]
Hi! Excellent job @johnnv1! There were typos/missing words in the card, so I took the liberty to rewrite some parts to make them easier to understand. Let me know if you are ok with the changes. Also, feel free to add some info under the `Who are the annotators?` section.

Additionally, I fixed the issue with streaming and renamed the `digits` feature to `objects`.

@lhoestq Are you ok with skipping the dummy data test here as it's tricky to generate it due to the splits separation logic?

File: datasets/ccagt/README.md [hunk 1]
```diff
@@ -0,0 +1,262 @@
...
+
+#### Who are the source language producers?
+
+The dataset was built using images from examinations (a gynecological exam, colposcopy and biopsy) of 15 women patients who were treated at the Gynecology and Colposcopy Outpatient Clinic of the [University Hospital Professor Polydoro Ernani de São Thiago of Federal University of Santa Catarina (HU-UFSC)](https://unihospital.ufsc.br/) and had 6 different diagnoses in their oncological exams. The samples were collected by the members of the Clinical Analyses Department: Ane Francyne Costa, Fabiana Botelho De Miranda Onofre, and Alexandre Sherlley Casimiro Onofre.
+
+### Annotations
+
+#### Annotation process
+
+The instances were annotated using the [labelbox](https://labelbox.com/) tool. The satellite category was labeled as a single dot, and the other categories were labeled as polygons. After the annotation process, all annotations was reviewed.
```
Comment (mariosasko) [inline datasets/ccagt/README.md:177]:
```suggestion
The instances were annotated using the [labelbox](https://labelbox.com/) tool. The satellite category was labeled as a single dot, and the other categories were labeled as polygons. After the annotation process, all annotations were reviewed.
```

File: datasets/ccagt/README.md [hunk 2]
```diff
@@ -0,0 +1,262 @@
...
+- `annotation`: A `PIL.Image.Image` object containing the annotation mask.
+
+#### `object detection`
+
+- `image`: A `PIL.Image.Image` object containing the image. Note that when accessing the image column: `dataset[0]["image"]` the image file is automatically decoded. Decoding of a large number of image files might take a significant amount of time. Thus it is important to first query the sample index before the `"image"` column, *i.e.* `dataset[0]["image"]` should **always** be preferred over `dataset["image"][0]`.
+- `objects`: a dictionary containing bounding boxes and labels of the cell objects 
+  - `bbox`: a list of bounding boxes (in the [coco](https://albumentations.ai/docs/getting_started/bounding_boxes_augmentation/#coco) format) corresponding to the objects present on the image
+  - `label`: a list of integers representing the category (7 categories to describe the objects in total; two to differentiate nucleolus organizer regions), with the possible values including `NUCLEUS` (1), `CLUSTER` (2), `SATELLITE` (3), `NUCLEUS_OUT_OF_FOCUS` (4), `OVERLAPPED_NUCLEI` (5), `NON_VIABLE_NUCLEUS` (6) and `LEUKOCYTE_NUCLEUS` (7).
```
Comment (mariosasko) [inline datasets/ccagt/README.md:116]: `datasets` / `transformers` expect the `0, ..., num_classes-1` label range

```suggestion
  - `label`: a list of integers representing the category (7 categories to describe the objects in total; two to differentiate nucleolus organizer regions), with the possible values including `NUCLEUS` (0), `CLUSTER` (1), `SATELLITE` (2), `NUCLEUS_OUT_OF_FOCUS` (3), `OVERLAPPED_NUCLEI` (4), `NON_VIABLE_NUCLEUS` (5) and `LEUKOCYTE_NUCLEUS` (6).
```

File: datasets/ccagt/ccagt.py [hunk 3]
```diff
@@ -0,0 +1,318 @@
...
+
+        def images_and_masks(basenames):
+            for bn in basenames:
+                yield build_path(bn), build_path(bn, "masks")
+
+        def images_and_boxes(basenames):
+            for bn in basenames:
+                image_id = self._bn_to_imageid[bn]
+                labels = [
+                    {"bbox": annotation["bbox"], "label": annotation["category_id"]}
```
Comment (mariosasko) [inline datasets/ccagt/ccagt.py:261]: To respect the `0, ..., num_classes-1` label range

```suggestion
                    {"bbox": annotation["bbox"], "label": annotation["category_id"] - 1}
```

File: datasets/ccagt/README.md [hunk 4]
```diff
@@ -0,0 +1,262 @@
...
+  - [Source Data](#source-data)
+  - [Annotations](#annotations)
+  - [Personal and Sensitive Information](#personal-and-sensitive-information)
+- [Considerations for Using the Data](#considerations-for-using-the-data)
+  - [Social Impact of Dataset](#social-impact-of-dataset)
+  - [Other Known Limitations](#other-known-limitations)
+- [Additional Information](#additional-information)
+  - [Dataset Curators](#dataset-curators)
+  - [Licensing Information](#licensing-information)
+  - [Citation Information](#citation-information)
```
Comment (mariosasko) [inline datasets/ccagt/README.md:47]:
```suggestion
## Table of Contents
- [Dataset Description](#dataset-description)
  - [Dataset Summary](#dataset-summary)
  - [Supported Tasks and Leaderboards](#supported-tasks-and-leaderboards)
  - [Languages](#languages)
- [Dataset Structure](#dataset-structure)
  - [Data Instances](#data-instances)
  - [Data Fields](#data-fields)
  - [Data Splits](#data-splits)
- [Dataset Creation](#dataset-creation)
  - [Curation Rationale](#curation-rationale)
  - [Source Data](#source-data)
  - [Annotations](#annotations)
  - [Personal and Sensitive Information](#personal-and-sensitive-information)
- [Considerations for Using the Data](#considerations-for-using-the-data)
  - [Social Impact of Dataset](#social-impact-of-dataset)
  - [Discussion of Biases](#discussion-of-biases)
  - [Other Known Limitations](#other-known-limitations)
- [Additional Information](#additional-information)
  - [Dataset Curators](#dataset-curators)
  - [Licensing Information](#licensing-information)
  - [Citation Information](#citation-information)
  - [Contributions](#contributions)
```

**Review by mariosasko — COMMENTED** [review]
Yes, feel free to add the instance segmentation config as described (+ update the task tags and `Supported tasks and leaderboards`).

I've left some suggestions:

File: datasets/ccagt/README.md (hunk 2 above)
Comment (johnnv1) [reply datasets/ccagt/README.md:116]: I think it makes more sense to keep the same order/IDs as in semantic segmentation, where 0 is for background/ignored pixels

**Comment by johnnv1** [comment]
Don't you think it makes sense to keep the same category IDs for all approaches? 

Now we have:
 - nucleus category ID equals 0 for object detection and instance segmentation
 - background category ID equals 0 (on the masks) for semantic segmentation

**Comment by mariosasko** [comment]
I find it weird to have a dummy label in object detection just to align the mapping with semantic segmentation. Instead, let's explain in the card (under Data Fields -> annotation) what the pixel values mean (background + object detection labels)

File: datasets/ccagt/README.md [hunk 5]
```diff
@@ -0,0 +1,273 @@
...
+
+### Supported Tasks and Leaderboards
+
+- `image-segmentation`: The dataset can be used to train a model for semantic segmentation or instance segmentation. Semantic segmentation consists in classifying each pixel of the image. Instance segmentation consists of doing object detection first and then using a semantic segmentation model inside detected objects. Success on this task is typically measured by achieving high values of [mean iou](https://huggingface.co/spaces/evaluate-metric/mean_iou) or [f-score](https://huggingface.co/spaces/evaluate-metric/f1) for pixels results. For instances results, this task is typically measured by achieving high values of [recall](https://huggingface.co/spaces/evaluate-metric/recall), [precision](https://huggingface.co/spaces/evaluate-metric/precision) and [f-score](https://huggingface.co/spaces/evaluate-metric/f1).
+  - Panoptic segmentation require an additional conversion of the original dataset. Panoptic segmentation is the combination of instance segmentation and semantic segmentation.
```
Comment (mariosasko) [inline datasets/ccagt/README.md:67]: Let's reorder the sentences here and remove the panoptic segmentation part to simplify the task entry:

```suggestion
- `image-segmentation`: The dataset can be used to train a model for semantic segmentation or instance segmentation. Semantic segmentation consists in classifying each pixel of the image. Success on this task is typically measured by achieving high values of [mean iou](https://huggingface.co/spaces/evaluate-metric/mean_iou) or [f-score](https://huggingface.co/spaces/evaluate-metric/f1) for pixels results. Instance segmentation consists of doing object detection first and then using a semantic segmentation model inside detected objects. For instances results, this task is typically measured by achieving high values of [recall](https://huggingface.co/spaces/evaluate-metric/recall), [precision](https://huggingface.co/spaces/evaluate-metric/precision) and [f-score](https://huggingface.co/spaces/evaluate-metric/f1).
```

File: datasets/ccagt/README.md [hunk 6]
```diff
@@ -0,0 +1,273 @@
...
+  'image': <PIL.JpegImagePlugin.JpegImageFile image mode=RGB size=1200x1600 at 0x276021C5EB8>,
+  'objects': {
+    'bbox': [
+      [36, 7, 13, 32],
+      [50, 7, 12, 32]
+    ], 
+    'label': [1, 5]
+  }
+```
+
```
Comment (mariosasko) [inline datasets/ccagt/README.md:104]: Let's add an example for the `instance segmentation`  config here.

File: datasets/ccagt/README.md [hunk 7]
```diff
@@ -0,0 +1,273 @@
...
+### Data Fields
+
+The data annotations have the following fields:
+
+#### `semantic segmentation` (default configuration)
+
+- `image`: A `PIL.Image.Image` object containing the image. Note that when accessing the image column: `dataset[0]["image"]` the image file is automatically decoded. Decoding of a large number of image files might take a significant amount of time. Thus it is important to first query the sample index before the `"image"` column, *i.e.* `dataset[0]["image"]` should **always** be preferred over `dataset["image"][0]`.
+- `annotation`: A `PIL.Image.Image` object containing the annotation mask. The mask has the category id values, with the possible values including `BACKGROUND` (0), `NUCLEUS` (1), `CLUSTER` (2), `SATELLITE` (3), `NUCLEUS_OUT_OF_FOCUS` (4), `OVERLAPPED_NUCLEI` (5), `NON_VIABLE_NUCLEUS` (6) and `LEUKOCYTE_NUCLEUS` (7).
```
Comment (mariosasko) [inline datasets/ccagt/README.md:112]:
```suggestion
- `annotation`: A `PIL.Image.Image` object containing the annotation mask. The mask has a single channel and the following pixel values are possible: `BACKGROUND` (0), `NUCLEUS` (1), `CLUSTER` (2), `SATELLITE` (3), `NUCLEUS_OUT_OF_FOCUS` (4), `OVERLAPPED_NUCLEI` (5), `NON_VIABLE_NUCLEUS` (6) and `LEUKOCYTE_NUCLEUS` (7).
```

**Review by mariosasko — COMMENTED** [review]
Nice! Three nits and we can merge:

File: datasets/ccagt/README.md [hunk 8]
```diff
@@ -0,0 +1,272 @@
...
+- `objects`: a dictionary containing bounding boxes and labels of the cell objects
+  - `bbox`: a list of bounding boxes (in the [coco](https://albumentations.ai/docs/getting_started/bounding_boxes_augmentation/#coco) format) corresponding to the objects present on the image
+  - `segment`: a list of segments in format of `[polygon_0, ..., polygon_n]`, where each polygon is `[x0, y0, ..., xn, yn]`.
+  - `label`: a list of integers representing the category (7 categories to describe the objects in total; two to differentiate nucleolus organizer regions), with the possible values including `NUCLEUS` (0), `CLUSTER` (1), `SATELLITE` (2), `NUCLEUS_OUT_OF_FOCUS` (3), `OVERLAPPED_NUCLEI` (4), `NON_VIABLE_NUCLEUS` (5) and `LEUKOCYTE_NUCLEUS` (6).
+
+
+### Data Splits
+
+The data is split randomly using the fixed seed into training, test and validation set. The training data contains 70% of the images and the testing and the validation data contain 15% of the images each. In total, the training set contains 6533 images and the testing and the validation set 1403 images each.
```
Comment (lhoestq) [inline datasets/ccagt/README.md:131]: Are the splits provided by the dataset authors ? Or is this an arbitrary splitting that you implemented yourself ?

We usually only use official splits, or we let the users define their own splits

Comment (johnnv1) [reply datasets/ccagt/README.md:131]: I am the author of the dataset, so yes it's the official splitting process 😄 

The difference to the official results (at [10.2139/ssrn.4126881](https://dx.doi.org/10.2139/ssrn.4126881)), was that in the paper we also added all the data from one slide to the test set and then performed this split... i can also add this slide to the test fold to be equals, but I would like to provide this to the users so that they can choose which slide they would like to separate at the start

**Review by mariosasko — APPROVED** [review]
All looks good now. Thanks!

PS: the CI failure is unrelated to the changes, so feel free to ignore it.

**Comment by johnnv1** [comment]
Ok, I can do that in the next few days. I will create a `lapix` organization, and I will add this dataset and also #4565

**Comment by johnnv1** [comment]
So, I think we can close this PR? I have already moved these files there.

The link of CCAgT dataset is: https://huggingface.co/datasets/lapix/CCAgT

🤗

**Comment by lhoestq** [comment]
Woohoo awesome !

closing this PR :)
